src/ray/common/scheduling/resource_instance_set.h

```diff
@@ -101,15 +101,75 @@ class NodeResourceInstanceSet {
+  /// and 1 placement group with 2 bundles. The bundle with index 1 contains 1 GPU and
+  /// the bundle with index 2 contains 2 GPU.
+  ///
+  /// The current node resource can be as follows:
+  /// resource id: total, available
+  /// GPU: [1, 1, 1], [0, 0, 0]
+  /// GPU_<pg_id>: [1, 1, 1], [1, 1, 1]
+  /// GPU_1_<pg_id>: [1, 0, 0], [1, 0, 0]
+  /// GPU_2_<pg_id>: [0, 1, 1], [0, 1, 1]
+  ///
+  /// Now, we want to allocate a task with 2 GPUs and in the placement group <pg_id>,
+  /// reflecting in the following resource demand:
+  /// GPU_<pg_id> : 2
+  ///
+  /// We will iterate though all the bundles in the placement group and bundle with
+  /// index=2 has the required capacity. So we will allocate the task to the 2 GPUs in
+  /// bundle 2 in placement group <pg_id> and the same allocation should be reflected in
+  /// the wildcard GPU resource. So the allocation will be:
+  /// GPU_<pg_id> : [0, 1, 1]
+  /// GPU_2_<pg_id> : [0, 1, 1]
+  ///
+  /// And as a result, after the allocation, current node resource will be:
+  /// resource id: total, available
+  /// GPU: [1, 1, 1], [0, 0, 0]
+  /// GPU_<pg_id>: [1, 1, 1], [1, 0, 0]
+  /// GPU_1_<pg_id>: [1, 0, 0], [1, 0, 0]
+  /// GPU_2_<pg_id>: [0, 1, 1], [0, 0, 0]
+  ///
   /// \param resource_id: The id of the resource to be allocated.
   /// \param demand: The resource amount to be allocated.
   ///
   /// \return the allocated instances, if allocation successful. Else, return nullopt.
   std::optional<std::vector<FixedPoint>> TryAllocate(ResourceID resource_id,
                                                      FixedPoint demand);
 
+  /// Allocate resource to the resource_id based on a provided reference allocation.
+  /// The function is used for placement group allocation. Making the allocation of
+  /// the wildcard resource be identical to the indexed resource allocation.
+  ///
+  /// The function assumes and also verifies that (1) the resource_id exists in the
+  /// node; (2) the available resources with resource_id on the node can satisfy the
+  /// provided ref_allocation.
+  ///
+  /// \param ref_allocation: The reference allocation used to allocate the resource_id
+  /// \param resource_id: The id of the resource to be allocated
+  void AllocateWithReference(const std::vector<FixedPoint> &ref_allocation,
+                             ResourceID resource_id);
+
   /// Map from the resource IDs to the resource instance values.
   absl::flat_hash_map<ResourceID, std::vector<FixedPoint>> resources_;
+
+  /// This is a derived map from the resources_ map. The map aggregates all the current
+  /// placement group indexed resources in resources_ by their original resource id and
+  /// pd id. The key of the map is the original resource id. The value is a map of pg id
+  /// to the corresponding placement group indexed resource ids. This map should always
+  /// be consistent with the _resource map.
```

Comment:
```suggestion
  /// be consistent with the resources_ map.
```

This is necessary for performance?

Reply:
Yes. Otherwise, whenever we do the allocation for placement group with only wildcard resource id specified, we will always need to iterate and parse the entries in the resource_ map and find all the corresponding indexed resources.